### PR TITLE
Support DDP in `PyTorch-Lightning`

### PR DESCRIPTION
**Pull request by Alnusjaponica**
## Motivation
Follow up #4322. 
Temporary, DDP is not supported in `PyTorchLightningPruningCallback` because of the problem described in #4322.
This PR make `PyTorchLightningPruningCallback` support DDP again.

## Description of the changes
This PR
- **Require users to call `callback.check_pruned()` in objective functions when they use DDP**
- Activates `test_pytorch_lightning_pruning_callback_ddp_monitor` and `test_pytorch_lightning_pruning_callback_ddp_unsupported_storage`
- Store intermediate values, pruning state and message directly in the storage

### CODE REVIEW

**Comment by Alnusjaponica** [comment]
I made this PR review ready. Note that this PR is based on #4322 and should be merged after it.

**Comment by toshihikoyanase** [comment]
@HideakiImamura This is a follow-up for #4322. Could you join the review, please?

**Comment by toshihikoyanase** [comment]
@Alnusjaponica  #4322 was merged into `master`. Could you rebase `master`, please?

**Review by toshihikoyanase — COMMENTED** [review]
Let me share my early comments.
I guess the change uses system attrs to store intermediate values even if users execute only a single process. I'm thinking of keeping them to simplify the logic. What do you think of it?

**Comment by Alnusjaponica** [comment]
@toshihikoyanase Thank you for your indication. I fixed the comment and am going to make change to use system_attr only under distributed situation.

File: optuna/integration/pytorch_lightning.py
Comment (Alnusjaponica) [reply optuna/integration/pytorch_lightning.py:13]: Thank you for your comment. I don't have any particular reason for this. Is it better to arrange them in a particular order (e.g. alphabetically)?

File: optuna/integration/pytorch_lightning.py
Comment (Alnusjaponica) [reply optuna/integration/pytorch_lightning.py:72]: I meant the latter one.

File: optuna/integration/pytorch_lightning.py
Comment (toshihikoyanase) [reply optuna/integration/pytorch_lightning.py:13]: > Is it better to arrange them in a particular order (e.g. alphabetically)

That makes sense. Could you arrange the keys in alphabetical order, please?

**Comment by toshihikoyanase** [comment]
@Alnusjaponica Thank you for your update. As we pair-programmed the code, we may simplify the code in terms of the following points:
- The empty intermediate values in system_attrs are only required for the DDP. We can skip it by moving the logic to `on_fit_start`
- The logic in `on_validation_end` is a bit complicated. We may be separate the logic for single-process optimization and DDP optimization
- `check_pruned` is only used for DDP optimization. It should be documented.
- `check_pruned` is assumed to be used with `_cachedStorage`. We may remove `#type ignored` if we assert it.
- The message can be removed from the system_attrs since the pruned epoch can generate it.

File: optuna/integration/pytorch_lightning.py
Comment (toshihikoyanase) [inline optuna/integration/pytorch_lightning.py:140]: A nit,
```suggestion
        Currently, ``intermediate_values`` are not properly propagated between processes due to
```

File: optuna/integration/pytorch_lightning.py
Comment (toshihikoyanase) [inline optuna/integration/pytorch_lightning.py:141]: `_CachedStorage` is a private class, so how about mentioning the storage cache without showing the class name?

```suggestion
       storage cache. Therefore, necessary information is kept in trial_system_attrs when the
```

File: optuna/integration/pytorch_lightning.py
Comment (toshihikoyanase) [inline optuna/integration/pytorch_lightning.py:153]: As talked offline, `intermediate_values` cannot be `None`.
So, how about adding an assertion instead of `type ignore` as follows?

```suggestion
        assert intermediate_values is not None
        for epoch, score in intermediate_values.items():
```

File: optuna/integration/pytorch_lightning.py
Comment (toshihikoyanase) [inline optuna/integration/pytorch_lightning.py:109]: Maybe we can simplify the code by removing the local variables.
```suggestion
        if not self.is_ddp_backend:
            self._trial.report(current_score.item(), step=epoch)
            if not self._trial.should_prune():
                return
            raise optuna.TrialPruned(f"Trial was pruned at epoch {epoch}.")
```

**Review by toshihikoyanase — COMMENTED** [review]
Thank you for your update. Let me add some small comments.

File: tests/integration_tests/test_pytorch_lightning.py
Comment (toshihikoyanase) [inline tests/integration_tests/test_pytorch_lightning.py:111]: In general, the official Python document recommends `Sequence` and `Mapping` for arguments instead of `List` and `Dict`. Can we use them here?

> To annotate arguments it is preferred to use an abstract collection type such as [Sequence](https://docs.python.org/3/library/typing.html#typing.Sequence) or [Iterable](https://docs.python.org/3/library/typing.html#typing.Iterable).

https://docs.python.org/3/library/typing.html#typing.List
https://docs.python.org/3/library/typing.html#typing.Dict

File: tests/integration_tests/test_pytorch_lightning.py
Comment (toshihikoyanase) [inline tests/integration_tests/test_pytorch_lightning.py:174]: A nit.
```suggestion
        # Evoke pruning manually.
``````suggestion
        # Evoke pruning manually.
```

File: tests/integration_tests/test_pytorch_lightning.py
Comment (toshihikoyanase) [inline tests/integration_tests/test_pytorch_lightning.py:212]: ditto.
```suggestion
        # Evoke pruning manually.
```

File: optuna/integration/pytorch_lightning.py
Comment (toshihikoyanase) [reply optuna/integration/pytorch_lightning.py:64]: Thanks!

File: optuna/integration/pytorch_lightning.py
Comment (toshihikoyanase) [inline optuna/integration/pytorch_lightning.py:153]: Maybe we can remove this logic and just sync the owned trial after #4431 is merged. How about adding TODO comment about it?

**Review by toshihikoyanase — COMMENTED** [review]
I confirmed that the updated callback worked with the PyTorch Lightning DDP example in optuna-examples. I have a small comment, but the change almost looks good to me.

<details>
<summary>Diff of pytorch_lightning_ddp.py</summary>
```diff
diff --git a/pytorch/pytorch_lightning_ddp.py b/pytorch/pytorch_lightning_ddp.py
index 030834d..6d609cb 100644
--- a/pytorch/pytorch_lightning_ddp.py
+++ b/pytorch/pytorch_lightning_ddp.py
@@ -78,6 +78,9 @@ class LightningNet(pl.LightningModule):
         self.log("val_acc", accuracy, sync_dist=True)
         self.log("hp_metric", accuracy, on_step=False, on_epoch=True, sync_dist=True)
 
+    def validation_epoch_end(self, output) -> None:
+        return
+
     def configure_optimizers(self) -> optim.Optimizer:
         return optim.Adam(self.model.parameters())
 
@@ -124,19 +127,22 @@ def objective(trial: optuna.trial.Trial) -> float:
     model = LightningNet(dropout, output_dims)
     datamodule = FashionMNISTDataModule(data_dir=DIR, batch_size=BATCHSIZE)
 
+    callback = PyTorchLightningPruningCallback(trial, monitor="val_acc")
     trainer = pl.Trainer(
         logger=True,
         limit_val_batches=PERCENT_VALID_EXAMPLES,
         enable_checkpointing=False,
         max_epochs=EPOCHS,
         gpus=-1 if torch.cuda.is_available() else None,
-        accelerator="ddp_cpu" if not torch.cuda.is_available() else None,
+        accelerator="cpu" if not torch.cuda.is_available() else None,
+        strategy="ddp_spawn",
         num_processes=os.cpu_count() if not torch.cuda.is_available() else None,
-        callbacks=[PyTorchLightningPruningCallback(trial, monitor="val_acc")],
+        callbacks=[callback],
     )
     hyperparameters = dict(n_layers=n_layers, dropout=dropout, output_dims=output_dims)
     trainer.logger.log_hyperparams(hyperparameters)
     trainer.fit(model, datamodule=datamodule)
+    callback.check_pruned()
 
     return trainer.callback_metrics["val_acc"].item()
```
</details>

File: optuna/integration/pytorch_lightning.py
Comment (Alnusjaponica) [reply optuna/integration/pytorch_lightning.py:156]: Thank you for your suggestion. I added TODO.

Comment (toshihikoyanase) [reply optuna/integration/pytorch_lightning.py:156]: Thanks! I'll approve this PR after all CI jobs are complete.

**Review by toshihikoyanase — APPROVED** [review]
We have some follow-up tasks as described in TODO comments, but I think we can work on them in a new PR.

LGTM. Thank you!

File: optuna/integration/pytorch_lightning.py
Comment (HideakiImamura) [inline optuna/integration/pytorch_lightning.py:84]: This setup only should be done in the rank zero process? In the following scenario, the value of the `_INTERMEDIATE_VALUE` is wrong.

```
------------------------> time
rank 0: (A) -> (B)
rank 1:                (A)
```
(A): call `on_fit_start` and init the `_INTERMEDIATE_VALUE`
(B): call `on_validation_end` and report the `_INTERMEDIATE_VALUE`
Actually, thanks to the specification of `on_fit_start` and `on_validation_end`, this problem might not occur, but I think it is safe to call this init only in the rank 0 process for future specification change of Pytorch Lightning.

File: optuna/integration/pytorch_lightning.py
Comment (HideakiImamura) [inline optuna/integration/pytorch_lightning.py:138]: This function is intended to be called by users by hand after the `Trainer.fit`. How about adding the concrete instruction on the document about where this function should be called?

File: optuna/integration/pytorch_lightning.py
Comment (HideakiImamura) [inline optuna/integration/pytorch_lightning.py:154]: If `self.is_ddp_backend = False`, this function raises an `AssertionError` here. How about early returning this function if `self.is_ddp_backend = False` to allow users to use same codes w/ or w/o DDP?

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR and sorry for the late reply. I checked the overall codes and basically looks good to me. Could you check my several comment?

File: optuna/integration/pytorch_lightning.py
Comment (Alnusjaponica) [reply optuna/integration/pytorch_lightning.py:195]: I added some explanation and example codes in docstrings.

File: optuna/integration/pytorch_lightning.py
Comment (Alnusjaponica) [reply optuna/integration/pytorch_lightning.py:215]: Thank you for your advice. I updated this method accordingly.

File: optuna/integration/pytorch_lightning.py
Comment (Alnusjaponica) [reply optuna/integration/pytorch_lightning.py:144]: Thank you for your careful review. I added an if-block so that `storage.set_trial_system_attr._INTERMEDIATE_VALUE` is initialized only from global zero.

File: optuna/integration/pytorch_lightning.py
Comment (Alnusjaponica) [inline optuna/integration/pytorch_lightning.py:158]:
```suggestion

assert isinstance(_study._storage, _CachedStorage)
```

**Review by HideakiImamura — APPROVED** [review]
Thanks for the update. LGTM.

**Pull request merged**
